pinot-core/src/main/java/org/apache/pinot/core/operator/query/SelectionPartiallyOrderedByDescOperation.java

```diff
@@ -0,0 +1,105 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.query;
+
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.pinot.common.request.context.ExpressionContext;
+import org.apache.pinot.core.common.BlockValSet;
+import org.apache.pinot.core.common.RowBasedBlockValueFetcher;
+import org.apache.pinot.core.operator.blocks.TransformBlock;
+import org.apache.pinot.core.operator.transform.TransformOperator;
+import org.apache.pinot.core.query.request.context.QueryContext;
+import org.apache.pinot.segment.spi.IndexSegment;
+
+
+/**
+ * An operator for order-by queries DESC that are partially sorted over the sorting keys.
+ *
+ * @see LinearSelectionOrderByOperator
+ */
+public class SelectionPartiallyOrderedByDescOperation extends LinearSelectionOrderByOperator {
+
+  private static final String EXPLAIN_NAME = "SELECT_PARTIAL_ORDER_BY_DESC";
+
+  private int _numDocsScanned = 0;
+  private long _numEntriesScannedPostFilter = 0;
+
+  public SelectionPartiallyOrderedByDescOperation(IndexSegment indexSegment, QueryContext queryContext,
+      List<ExpressionContext> expressions, TransformOperator transformOperator, int sortedExpr) {
+    super(indexSegment, queryContext, expressions, transformOperator, sortedExpr);
+    assert queryContext.getOrderByExpressions() != null;
+    Preconditions.checkArgument(queryContext.getOrderByExpressions().stream()
+            .filter(expr -> expr.getExpression().getType() == ExpressionContext.Type.IDENTIFIER)
+            .findFirst()
+            .orElseThrow(() -> new IllegalArgumentException("The query is not order by identifiers"))
+            .isDesc(),
+        "%s can only be used when the first column in order by is DESC", EXPLAIN_NAME);
+  }
+
+  @Override
+  protected void fetch(ListBuilder listBuilder) {
+
+    List<TransformBlock> blocks = new ArrayList<>();
+
+    TransformBlock transformBlock;
+    int numColumnsProjected = _transformOperator.getNumColumnsProjected();
+    try {
+      while ((transformBlock = _transformOperator.nextBlock()) != null) {
+        int numDocsFetched = transformBlock.getNumDocs();
+        _numDocsScanned += numDocsFetched;
+        blocks.add(transformBlock);
```

Comment:
(MAJOR) We cannot store all blocks here because the `ProjectionBlock` shares the same `DataBlockCache` underlying, where the previous value is overridden.
For descending order, the best we can do currently is to iterate in reverse order from each block, and keep up to `LIMIT` rows per block. When getting the next block, if we gathered enough rows, we can throw away the rows from the previous block.

Reply:
That is sad. In that case it may not make sense to apply this optimization. Should we wait to have descending iterators/cursors?

Reply:
I've just changed the code so it fetchs the last <limit> rows of each block.